include/nbl/builtin/hlsl/emulated/float64_t.hlsl

```diff
@@ -0,0 +1,657 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_HLSL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_HLSL_INCLUDED_
+
+#include <nbl/builtin/hlsl/emulated/float64_t_impl.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+    /*enum E_ROUNDING_MODE
+    {
+        FLOAT_ROUND_NEAREST_EVEN,
+        FLOAT_ROUND_TO_ZERO,
+        FLOAT_ROUND_DOWN,
+        FLOAT_ROUND_UP
+    };*/
+
+    // currently only FLOAT_ROUND_TO_ZERO is supported, cannot implement partial specialization in this case due to dxc bug https://github.com/microsoft/DirectXShaderCompiler/issues/5563
+    // TODO: partial specializations with new template parameter `E_ROUNDING_MODE RoundingMode`
+    template<bool FastMath = true, bool FlushDenormToZero = true>
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+        using this_t = emulated_float64_t<FastMath, FlushDenormToZero>;
+
+        storage_t data;
+
+        // constructors
+        /*static emulated_float64_t create(uint16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(this_t val)
+        {
+            return val;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int32_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(int64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int64_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint32_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(uint64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint64_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float32_t val)
+        {
+            this_t output;
+            output.data = emulated_float64_t_impl::castFloat32ToStorageType<FlushDenormToZero>(val);
+            return output;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float64_t val)
+        {
+#ifdef __HLSL_VERSION
+            emulated_float64_t retval;
+            uint32_t lo, hi;
+            asuint(val, lo, hi);
+            retval.data = (uint64_t(hi) << 32) | uint64_t(lo);
+            return retval;
```

Comment:
this is because `bit_cast` isnt treated as constexpr and `asuint` is?

Reply:
`bit_cast` doesn't lead to emission of `Float64` capability, even with Optimization level 0

https://godbolt.nsc.devsh.eu/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,selection:(endColumn:31,endLineNumber:2,positionColumn:31,positionLineNumber:2,selectionStartColumn:31,selectionStartLineNumber:2,startColumn:31,startLineNumber:2),source:'%23include+%22nbl/builtin/hlsl/cpp_compat.hlsl%22%0A%23include+%22nbl/builtin/hlsl/bit.hlsl%22%0A%0A%5Bnumthreads(8,8,8)%5D%0Avoid+main(uint32_t3+gl_GlobalInvocationID+:+SV_DispatchThreadID)%0A%7B%0A++++static+const+double+PI+%3D+3.14159%3B%0A++++vk::RawBufferStore%3Cuint64_t%3E(0xdeadbeefBADC0FFEull,nbl::hlsl::bit_cast%3Cuint64_t%3E(PI))%3B%0A%7D%0A'),l:'5',n:'0',o:'HLSL+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:nsc_release_upstream,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,libs:!(),options:'-T+cs_6_7+-E+main+-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+NSC+(Release)+(Editor+%231)',t:'0')),k:50,l:'4',m:35.90604026845637,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'NSC+(Release)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+NSC+(Release)+(Compiler+%231)',t:'0')),header:(),l:'4',m:64.09395973154362,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

Reply:
Bitcast is safe with constexpr, we already went through this with @Przemog1 . It was probably an oversight